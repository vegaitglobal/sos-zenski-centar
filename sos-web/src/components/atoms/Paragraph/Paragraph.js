import { forwardRef } from 'react';
import { StyledParagraph } from './Paragraph.styles';

export const Paragraph = forwardRef(
  (
    {
      type,
      as = 'p',
      fontWeight,
      textTransform,
      textAlign,
      color,
      children,
      ...props
    },
    ref,
  ) => (
    <StyledParagraph
      ref={ref}
      as={as}
      $type={type}
      $fontWeight={fontWeight}
      $textTransform={textTransform}
      $textAlign={textAlign}
      $color={color}
      {...props}
    >
      {children}
    </StyledParagraph>
  ),
);
