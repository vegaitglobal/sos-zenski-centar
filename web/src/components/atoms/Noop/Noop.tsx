import { ReactElement } from 'react';

export const Noop = (): ReactElement => <></>;
