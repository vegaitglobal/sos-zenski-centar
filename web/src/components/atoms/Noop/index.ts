export { Noop as default } from './Noop';
