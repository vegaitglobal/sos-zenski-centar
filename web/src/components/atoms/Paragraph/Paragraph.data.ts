import type { ReactNode } from 'react';
import { ColorType } from '../../../styles/config/theme';
import type { FontWeight, TextAlign, TextTransform } from '../../../styles/config/variables';

export enum ParagraphType {
	Large = 'large',
	Normal = 'normal',
	Small = 'small',
}

export interface ParagraphProps {
	as?: string | React.ComponentType<unknown>;
	type?: ParagraphType;
	fontWeight?: FontWeight;
	textTransform?: TextTransform;
	textAlign?: TextAlign;
	color?: ColorType;
	children: ReactNode;
}

export interface StyledParagraphProps {
	$type: ParagraphType;
	$fontWeight?: FontWeight;
	$textTransform?: TextTransform;
	$textAlign?: TextAlign;
	$color?: ColorType;
}
