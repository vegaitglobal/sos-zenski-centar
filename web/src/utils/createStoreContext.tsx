import { createContext, PropsWithChildren, FunctionComponent, Context, useContext } from 'react';
import useStoreFactory from '../hooks/useStoreFactory';

/**
 * Creates a context for the store created by the passed `storeFactory`.
 * Additionally returns the `Provider` component and the hook for the created store's context.
 *
 * @param name - name of the store
 * @param storeFactory - a function that returns an object representing the store
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const createStoreContext = <TStore extends Record<string, unknown>>(
	name: string,
	storeFactory: () => TStore,
): [FunctionComponent<unknown>, () => TStore, Context<TStore>] => {
	const context = createContext<TStore>({} as TStore); // the initial value is only to satisfy TS and won't be used

	/**
	 * Initializes and contains the context for the `TStore`.
	 */
	function StoreProvider({ children }: PropsWithChildren<unknown>): JSX.Element {
		const store = useStoreFactory(storeFactory);
		return <context.Provider value={store}>{children}</context.Provider>;
	}

	const useStore = (): TStore => {
		const store = useContext(context);
		if (store) {
			return store;
		}
		throw new ReferenceError(
			`${name} is not initialized, make sure that your component is wrapped in the 'Provider' of the store.`,
		);
	};

	return [StoreProvider, useStore, context];
};

export default createStoreContext;
