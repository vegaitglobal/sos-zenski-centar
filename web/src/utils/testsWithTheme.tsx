import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { ReactElement, ComponentType, ReactNode } from 'react';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType = ({ children }) => componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
