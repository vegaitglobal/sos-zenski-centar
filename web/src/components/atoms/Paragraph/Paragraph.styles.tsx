import styled from 'styled-components';
import { typeStyle } from '../../../styles/config/typeStyles';
import { StyledParagraphProps } from './Paragraph.data';

export const StyledParagraph = styled.p<StyledParagraphProps>`
	${({ $type }) => typeStyle[$type]};
	${({ $fontWeight }) => $fontWeight && `font-weight:  ${$fontWeight}`};
	${({ $textTransform }) => $textTransform && `text-transform:  ${$textTransform}`};
	${({ $textAlign }) => $textAlign && `text-align: ${$textAlign}`};
	${({ theme, $color }) => $color && `color:  ${theme.color[$color]}`};
`;
