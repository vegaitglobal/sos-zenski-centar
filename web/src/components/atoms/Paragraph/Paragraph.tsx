import { forwardRef } from 'react';
import { ParagraphProps, ParagraphType } from './Paragraph.data';
import { StyledParagraph } from './Paragraph.styles';

export const Paragraph = forwardRef<HTMLParagraphElement, ParagraphProps>(
	(
		{
			type = ParagraphType.Normal,
			as = 'p',
			fontWeight,
			textTransform,
			textAlign,
			color,
			children,
			...props
		},
		ref,
	) => (
		<StyledParagraph
			data-testid="Paragraph"
			ref={ref}
			as={as}
			$type={type}
			$fontWeight={fontWeight}
			$textTransform={textTransform}
			$textAlign={textAlign}
			$color={color}
			{...props}
		>
			{children}
		</StyledParagraph>
	),
);
