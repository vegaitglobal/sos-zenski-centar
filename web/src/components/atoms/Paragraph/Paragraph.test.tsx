import renderer from 'react-test-renderer';
import Noop from '../Noop';
import Paragraph from './index';
import { renderWithTheme } from '../../../utils/testsWithTheme';

describe('Paragraph', () => {
	describe('Render', () => {
		it('Renders Paragraph', () => {
			const { getByTestId } = renderWithTheme(
				<Paragraph>
					<Noop />
				</Paragraph>,
			);

			const element = getByTestId('Paragraph');

			expect(element).toBeTruthy();
		});

		it('Renders as span tag', () => {
			const { getByTestId } = renderWithTheme(
				<Paragraph as="span">
					<Noop />
				</Paragraph>,
			);

			const element = getByTestId('Paragraph');

			expect(element.tagName).toBe('SPAN');
		});

		it('Renders correctly', () => {
			const tree = renderer
				.create(
					<Paragraph>
						<Noop />
					</Paragraph>,
				)
				.toJSON();

			expect(tree).toMatchSnapshot();
		});
	});
	describe('Content', () => {
		it('Has valid content', () => {
			const { getByTestId } = renderWithTheme(<Paragraph>Paragraph text</Paragraph>);

			const element = getByTestId('Paragraph');

			expect(element).toHaveTextContent('Paragraph text');
		});

		it('Can have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Paragraph>
					<span>P</span>aragraph
				</Paragraph>,
			);

			const element = getByTestId('Paragraph');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});
});
