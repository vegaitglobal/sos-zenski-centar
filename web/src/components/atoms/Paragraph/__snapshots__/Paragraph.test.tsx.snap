// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph Render Renders correctly 1`] = `
<p
  className="sc-bdnxRM jcQGal"
  data-testid="Paragraph"
/>
`;
